### DISCUSSION TITLE
CosBox Version 1

### DISCUSSION THREAD

**Discussion by cmcooper1971** [pschatzmann/arduino-audio-tools#1629] [Show and tell]
Hi All,

I wanted to share with everyone who uses Phils Audio Tools library my project, which I have built for a friend and his Star Wars community. You can see what they do at [https://www.youtube.com/@norwichdroids](https://www.youtube.com/@norwichdroids) and to say thanks to Phil Schatzmann before I post on other platforms.

![Built Unit](https://github.com/user-attachments/assets/78d1cb92-439b-4912-9e16-58c5afe62e85)![Transmitter PCB](https://github.com/user-attachments/assets/cf435ed5-833c-47a5-b244-7c36ff830f6e)
![Receiver PCB](https://github.com/user-attachments/assets/9896f1c2-eb7d-42a2-ba05-48eda7c3c227)

It’s called CosBox and is a Class D I2S Amplifier, using a Texas Instruments TAS5760L I2S IC, MP3 Player with SD Card, I2S Microphone input for use as a megaphone with Voice Changer. All controlled by an ESP32 WROOM. 

Why include a Wi-Fi unit I hear you asking?!

The answer is simple, the receiver is accompanied by a transmitter, which the user uses, to trigger chosen tracks in certain orders, control the volume and to switch between the MP3 player and the Microphone input. This is achieved through traditional push switches sown into a fingerless glove, plus single and double tap detection, as the transmitter is complete with an MPU. In a later release, I will use the MPU for full gesture control, to control the receiver unit in new ways.

The user will be able to wear their chosen costume, select from a profile of pre-stored characters and the sound tracks from the various films. 

The unit is powered by an Lithium 18650 battery, and includes full battery charging and battery protection circuitry, 3.3V regulator circuit for the ESP32 and associated Digital ICs, plus a 12V boost convertor for the main Amplifier power.

I wanted to say thanks to Phil for his work as without his library, I would have struggled to get this project off the ground, not alone completed.

When I have videos of the characters in full costume dress, using the platform, I will share them.

Thanks Phil.

**Comment by pschatzmann** [comment]
Cool: looking forward to see the videos...

**Comment by peteDDD** [comment]
can you share the code you used for selecting the sounds to be played?  Do you experience any slight background noise when sounds are not being played?

**Reply by cmcooper1971** [reply]
Hi Pete,

Your first question, the tracks are triggered by a transmitter unit I built, which sends and integer to the receiver, which then plays the chosen track. Although this varies per button, as some buttons simply play the next track, while others plays a specified track. So I'm not sure my code will help. Is there anything in particular that you are after and I may be able to help?

For your second question, yes, I did get noise when tracks were not playing and Phil helped solve this by using one of the functions to see if there were any bytes remaining and if there were not, I simply disabled the I2S streams.

This is the part of the loop which plays either the MP3 or WAV file:

```
	if (playerControlFlag == true && micControlFlag == false) {

		size_t copied = mp3.copy();			// Start copy and check data byte size

		if (copied == 0 && playAudioFile == true) {

			// No more data to copy, end of file reached

			debugln("End of file reached.");

			disablePlayer();
			audioFile.close();
			playAudioFile = false;

		}

	}
```

This is the function for disabling all of my I2S streams:

```
// Disable I2S Streams

void disablePlayer() {

	mic.end();
	decoderWav.end();
	decoderMP3.end();
	mp3.end();
	in.end();
	out.end();

	digitalWrite(amberLED, false);

} // Close function
```

These are my two functions for playing WAV or MP3 files:


```
// Enable Player & disable MIC

void enablePlayerMP3() {

	mic.end();
	in.end();
	out.end();

	// Start I2S out

	AudioLogger::instance().begin(Serial, AudioLogger::Warning);

	auto config_out = out.defaultConfig(TX_MODE);
	config_out.copyFrom(info);
	config_out.i2s_format = I2S_STD_FORMAT;
	config_out.is_master = true;
	config_out.port_no = 1;
	config_out.pin_bck = I2S_BCLK_1;
	config_out.pin_ws = I2S_LRC_1;
	config_out.pin_data = I2S_DOUT_1;
	config_out.pin_mck = I2S_MCLK_1;
	out.begin(config_out);

	// Start player

	decoderMP3.begin();
	mp3.begin(decoderMP3, audioFile);

	debugln("MP3 Player Enabled");
	digitalWrite(amberLED, HIGH);

	playerControlFlagChanged = false;

	displayVolumeRegister(false);
	displayFaultRegister(false);

}	// Close function

```

```
// Enable Player & disable MIC

void enablePlayerWAV() {

	mic.end();
	in.end();
	out.end();

	// Start I2S out

	AudioLogger::instance().begin(Serial, AudioLogger::Warning);

	auto config_out = out.defaultConfig(TX_MODE);
	config_out.copyFrom(info);
	config_out.i2s_format = I2S_STD_FORMAT;
	config_out.is_master = true;
	config_out.port_no = 1;
	config_out.pin_bck = I2S_BCLK_1;
	config_out.pin_ws = I2S_LRC_1;
	config_out.pin_data = I2S_DOUT_1;
	config_out.pin_mck = I2S_MCLK_1;
	out.begin(config_out);

	// Start player

	decoderWav.begin();
	mp3.begin(decoderWav, audioFile);

	debugln("WAV Player Enabled");
	digitalWrite(amberLED, HIGH);

	playerControlFlagChanged = false;

	displayVolumeRegister(false);
	displayFaultRegister(false);

}	// Close function

```


You will see that the player is disabled along with closing the audio file. This all works seamlessly and I am very happy with the result.

Regards,

Christopher

**Comment by peteDDD** [comment]
Thanks @cmcooper1971 .

We ended up in similar places...   My demo code is below.
I tried to out.end() my stream to kill the hum coming out through the DAC but I found that ended up somehow putting the decoding in a confused state and all wavs after the first one were garbled due to some of the decoding setting going out the window.   I found that disabling the DAC channel worked perfectly, was easy, and I did not risk any memory leaks (if there are any) expressing themselves by opening and closing objects in the stream.

The demo below steps plays three wavs by stepping through two wav files (file1, file2, file1) so I could prove to myself that there was no need to close or "rewind" the wav files.  My real app will use the playWAV and checkIfWavIsDonePlaying functions with playWAV being passed the appropriate filenames or index to the appropriate file name.

Thanks again.
Pete

```
#include <Arduino.h>

#include "AudioTools.h"
#include "AudioTools/AudioLibs/AudioSourceSPIFFS.h"
#include "AudioTools/AudioCodecs/CodecWAV.h"

#include "driver/dac.h"

// list of wav files to play, saved in data folder
const char *wavFiles[] = {"/earth_move.wav", "/tremble.wav", "/earth_move.wav"};
const int numOfWavFiles = sizeof(wavFiles) / sizeof(wavFiles[0]);
uint8_t currentWavFileIndex = 0;
bool b_AskedToPlayWav = false;


// StreamCopy audioFile -> EncodedAudioStream decoder -> VolumeStream volume -> DACout
AnalogAudioStream DACout;
VolumeStream volume(DACout);
EncodedAudioStream decoder(&volume, new WAVDecoder()); // Decoding stream
StreamCopy copier;
File audioFile;


void playWAV(const char *fileName)
{
  audioFile = SPIFFS.open(fileName, "r");
  b_AskedToPlayWav = true;
  dac_output_enable(DAC_CHANNEL_2);
  copier.begin(decoder, audioFile);
}

bool checkIfWavIsDonePlaying()
{
  if (!copier.copy())
  {
    b_AskedToPlayWav = false; // done playing wave
    dac_output_disable(DAC_CHANNEL_2); // silence the DAC output
    return true;              // done playing wav
  }
  else
  {
    return false; // still playing wav
  }
}

void setup()
{
  Serial.begin(115200);
  SPIFFS.begin();

  // setup output
  auto DACoutCfg = DACout.defaultConfig();
  DACout.begin(DACoutCfg);

  // set initial volume  (ref: https://github.com/pschatzmann/arduino-audio-tools/wiki/Volume-Control)
  volume.begin(DACoutCfg); // we need to provide the bits_per_sample and channels
  volume.setVolume(0.5); // accepts floatvalues0.0 to 1.0

  decoder.begin();

  playWAV(wavFiles[currentWavFileIndex]);
}

void loop()
{
  if (b_AskedToPlayWav)
  {
    if (checkIfWavIsDonePlaying())
    {
      if (currentWavFileIndex++ < numOfWavFiles - 1)
      {
        Serial.println("PLAY NEXT FILE");  // shows when we actually start the playWAV call
        playWAV(wavFiles[currentWavFileIndex]); // play the next WAV file
      }
    }
  }

  // do other stuff here...
}
```

**Reply by pschatzmann** [reply]
I see the issue that you don't start the decoder when you change the file: so the codec does not know about this and it will interpret the wav header as audio! It would be most efficient if you would call begin on the WAVDecoder but since you don't store this as object you can do this on the EncodedAudioStream as well.

ps. don't forget to close your files!

**Reply by peteDDD** [reply]
Thanks!

So, we will go with this for the two functions.  Free resources and perfectly quiet DAC output after playing WAV.   Now I just need to make my wav files cleaner...

```
void playWAV(const char *fileName)
{
  audioFile = SPIFFS.open(fileName, "r");
  b_AskedToPlayWav = true;
  DACout.begin();
  decoder.begin();
  copier.begin(decoder, audioFile);
}

bool checkIfWavIsDonePlaying()
{
  if (!copier.copy())
  {
    audioFile.close();
    b_AskedToPlayWav = false; // done playing wave
    decoder.end();
    DACout.end();
    return true;              // done playing wav
  }
  else
  {
    return false; // still playing wav
  }
}
```
